File: LivrusVirtus/wxWindows/wx.NET/VCPP.NET/Cadastros.cpp
#include "stdafx.h"
#include "Cadastros.h"
#include "Resource.h"
#include "RotinasGlobais.h"

namespace Livrus
{
  frmCadastros::frmCadastros( wx::Window* parent, int nFiltroCodigo, 
    Boolean bIniciaBotoes ): wx::Panel(parent, -1, 
    System::Drawing::Point(0, 0), System::Drawing::Size(330, 140)) {

    labCodigo = new wx::StaticText(this, (int)idt::ID_FRMCADASTROS_LABCODIGO,
      "Código:", System::Drawing::Point(8, 8), System::Drawing::Size(36, 13), 0);          
    edtCodigo = new TextValidatorCtrl(this, (int)idt::ID_FRMCADASTROS_EDTCODIGO,
      "", System::Drawing::Point(8, 24), System::Drawing::Size(121, 21), 0, 
      nFiltroCodigo);
    btnNovo = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNNOVO, 
    	"&Novo", System::Drawing::Point(144, 24), System::Drawing::Size(75, 25), 0);
    labDescricao = new wx::StaticText(this, (int)idt::ID_FRMCADASTROS_LABDESCRICAO,
      "Descrição:", System::Drawing::Point(8, 48), System::Drawing::Size(51, 13), 0);  
    edtDescricao = new wx::TextCtrl(this, (int)idt::ID_FRMCADASTROS_EDTDESCRICAO,
      "", System::Drawing::Point(8, 64), System::Drawing::Size(217, 21), 0);  
  
    if (bIniciaBotoes) {
      IniciaBotoes();
      FazLayout();
    }
    edtCodigo->MaxLength = 10;
    edtDescricao->MaxLength = 30;
    Registros = new ArrayList();
    nLinhaRegistro = 0;
  }

  // Necessário por causa da tabulação dos campos
  Void frmCadastros::IniciaBotoes() {
    btnPesquisar = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNPESQUISAR,
      "&Pesquisar", System::Drawing::Point(248, 8), System::Drawing::Size(75, 25), 0);  
    btnAnterior = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNANTERIOR,
      "&Anterior", System::Drawing::Point(248, 40), System::Drawing::Size(75, 25), 0);  
    btnProximo = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNPROXIMO,
      "Próxi&mo", System::Drawing::Point(248, 72), System::Drawing::Size(75, 25), 0);  
    btnSalvar = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNSALVAR,
      "&Salvar", System::Drawing::Point(8, 104), System::Drawing::Size(75, 25), 0);  
    btnExcluir = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNEXCLUIR,
      "&Excluir", System::Drawing::Point(88, 104), System::Drawing::Size(75, 25), 0);  
    btnLimpar = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNLIMPAR,
      "&Limpar", System::Drawing::Point(168, 104), System::Drawing::Size(75, 25), 0); 
    btnFechar = new wx::Button(this, (int)idt::ID_FRMCADASTROS_BTNFECHAR,
      "&Fechar", System::Drawing::Point(248, 104), System::Drawing::Size(75, 25), 0);  
  
    btnLimparClick();
  }
  
  Void frmCadastros::FazLayout(){
    wx::BoxSizer* sizer_1 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::BoxSizer* sizer_2 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::BoxSizer* sizer_3 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::BoxSizer* sizer_4 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::BoxSizer* sizer_5 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::GridSizer* grid_sizer_2 = new wx::GridSizer(2, 1, 0, 0);
    wx::GridSizer* grid_sizer_1 = new wx::GridSizer(2, 2, 0, 1);
    grid_sizer_1->Add(labCodigo, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_1->Add(20, 10, 0, 0, 0);
    grid_sizer_1->Add(edtCodigo, 0, 0, 0);
    grid_sizer_1->Add(btnNovo, 0, wx::Direction::wxLEFT, 6);
    sizer_5->Add(grid_sizer_1, 0, 0, 0);
    grid_sizer_2->Add(labDescricao, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_2->Add(edtDescricao, 0, 0, 2);
    sizer_5->Add(grid_sizer_2, 0, 0, 0);
    sizer_3->Add(sizer_5, 1, 0, 0);
    sizer_4->Add(btnPesquisar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_4->Add(btnAnterior, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_4->Add(btnProximo, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_3->Add(sizer_4, 0, wx::Alignment::wxALIGN_RIGHT, 0);
    sizer_1->Add(sizer_3, 1, wx::Direction::wxALL|wx::Stretch::wxEXPAND, 4);
    sizer_2->Add(btnSalvar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnExcluir, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnLimpar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnFechar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_1->Add(sizer_2, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 4);
    AutoLayout = true;
    SetSizer(sizer_1);
    sizer_1->Fit(this);
    sizer_1->SetSizeHints(this);
  }

	Void frmCadastros::InformaLimpaDados(Boolean bInformar){
		if (bInformar) {
      Campos = dynamic_cast<ArrayList*>(Registros->Item[nLinhaRegistro]);
		  edtCodigo->Value=Campos->Item[0]->ToString();
		  edtDescricao->Value=Campos->Item[1]->ToString();
		} else {
		  edtCodigo->Clear();
		  edtDescricao->Clear();
		  edtCodigo->Enabled = true;
		  edtCodigo->SetFocus();
		}
	}

	Void frmCadastros::HabilitaDados(Boolean bHabilita){
		edtDescricao->Enabled = bHabilita;
	}

	Void frmCadastros::btnNovoClick(){
		btnSalvar->Enabled = true;
		btnExcluir->Enabled = false;
		btnPesquisar->Enabled = false;
		btnAnterior->Enabled = false;
		btnProximo->Enabled = false;
		btnNovo->Enabled = false;

    nLinhaRegistro = 0;
		InformaLimpaDados(false);
		HabilitaDados(true);
	}

	Void frmCadastros::NovoDado(String* sTextoSql){
    if (RotinasGlobais::Rotinas->ConsultaDadosLista(Registros, sTextoSql)) {
      Campos = dynamic_cast<ArrayList*>(Registros->Item[0]);
      edtCodigo->Value = Campos->Item[0]->ToString();
      edtCodigo->Enabled = false;
      edtDescricao->SetFocus();
    }
	}

	Void frmCadastros::btnLimparClick(){
		btnSalvar->Enabled = false;
		btnExcluir->Enabled = false;
		btnPesquisar->Enabled = true;
		btnAnterior->Enabled = false;
		btnProximo->Enabled = false;
		btnNovo->Enabled = true;

    nLinhaRegistro = 0;
		InformaLimpaDados(false);
		HabilitaDados(false);
	}

	Boolean frmCadastros::ExcluirDados(Boolean bValidaDados, String* sTextoSql){
    if (bValidaDados) {      
      if ( wx::MessageDialog::MessageBox(Resource::STR_CERTEXCL, "Confirme",
       wx::Dialog::wxICON_QUESTION | wx::Dialog::wxYES_NO) == wx::Dialog::wxYES ) {
        if (RotinasGlobais::Rotinas->AtualizaDados(Arquiva, sTextoSql)) {
           btnLimparClick();
           return true;
        }      
      }
    }
    return false;
	}

	Boolean frmCadastros::PesquisarDados(String* sTextoSql, String* sMensagemErro){
		if (edtCodigo->Value->Length==0) {
		  btnAnterior->Enabled = true;
		  btnProximo->Enabled = true;
		}

		if (RotinasGlobais::Rotinas->ConsultaDadosLista(Registros,sTextoSql)) {
			 InformaLimpaDados(true);
			 HabilitaDados(true);

			 btnSalvar->Enabled = true;
			 btnExcluir->Enabled = true;
			 btnPesquisar->Enabled = false;
       return true;
		} else
		  wx::MessageDialog::MessageBox(sMensagemErro, "Erro", 
        wx::Dialog::wxOK | wx::Dialog::wxICON_ERROR);

    return false;
	}

	Boolean frmCadastros::registroAnterior(){
		if (nLinhaRegistro > 0) {
			nLinhaRegistro=nLinhaRegistro - 1;
			InformaLimpaDados(true);
      return true;
    } else
      return false;
	}

	Boolean frmCadastros::registroProximo(){
		if (nLinhaRegistro < (Registros->Count-1)) {
			nLinhaRegistro=nLinhaRegistro + 1;
			InformaLimpaDados(true);
      return true;
    } else
      return false;
	}

	Boolean frmCadastros::SalvarDados(Boolean bValidaDados, String* sTextoSql,
    String* sTextoUpdate, String* sTextoInsert, Boolean bLimparDados){
    OleDbCommand* Consulta = new OleDbCommand();
    String* sArquiva;
  
  	if (bValidaDados) {    
  	   if (RotinasGlobais::Rotinas->ConsultaDados(Consulta,sTextoSql))
         sArquiva = sTextoUpdate;
       else
    	   sArquiva = sTextoInsert;	  	     
  
       if (RotinasGlobais::Rotinas->AtualizaDados(Arquiva, sArquiva)) {
    	   if (bLimparDados)
           btnLimparClick();
         return true;
       }
  	}
    return false;
	}
}

File: LivrusVirtus/wxWindows/wxRuby/Cadastros.rb
require 'wxruby'
require 'TextValidatorCtrl'
require 'ListaStr'
require 'resource'
require 'RotinasGlobais'

class FrmCadastros < Wx::Panel
  ID_FRMCADASTROS_BTNANTERIOR, ID_FRMCADASTROS_BTNEXCLUIR, 
  ID_FRMCADASTROS_BTNFECHAR, ID_FRMCADASTROS_BTNLIMPAR, 
  ID_FRMCADASTROS_BTNNOVO, ID_FRMCADASTROS_BTNPESQUISAR, 
  ID_FRMCADASTROS_BTNPROXIMO, ID_FRMCADASTROS_BTNSALVAR, 
  ID_FRMCADASTROS_EDTCODIGO, ID_FRMCADASTROS_EDTDESCRICAO, 
  ID_FRMCADASTROS_LABCODIGO, ID_FRMCADASTROS_LABDESCRICAO =
  3000,3001,3002,3003,3004,3005,3006,3007,3008,3009,3010,3011,3012

  def initialize(parent, nFiltroCodigo, bIniciaBotoes)
    super(parent, -1, Wx::Point.new(398, 295), 
      Wx::Size.new(338, 163), Wx::TAB_TRAVERSAL,"") 
    set_client_size(Wx::Size.new(330, 136))
    evt_close {|event| on_close(event)}

	  @labCodigo = Wx::StaticText.new(self, ID_FRMCADASTROS_LABCODIGO, 
      "Código:", Wx::Point.new(8, 8), Wx::Size.new(36, 13) )
  	@edtCodigo = TextValidatorCtrl.new(self, ID_FRMCADASTROS_EDTCODIGO, "", 
      Wx::Point.new(8, 24), Wx::Size.new(121, 21), 0, nFiltroCodigo )

    @edtCodigo.set_max_length(10)
		@btnNovo = Wx::Button.new(self, ID_FRMCADASTROS_BTNNOVO, "&Novo",
      Wx::Point.new(144, 24), Wx::Size.new(75, 25) )
	  @labDescricao = Wx::StaticText.new(self, ID_FRMCADASTROS_LABDESCRICAO, 
      "Descrição:", Wx::Point.new(8, 48), Wx::Size.new(51, 13) )
	  @edtDescricao = Wx::TextCtrl.new(self, ID_FRMCADASTROS_EDTDESCRICAO, "", 
      Wx::Point.new(8, 64), Wx::Size.new(217, 21) )
    @edtDescricao.enable(false)
    @edtDescricao.set_max_length(30)

    if bIniciaBotoes then
      IniciaBotoes()
      FazLayout()
    end

    @lstRegistros = LISTAMSTR.new
    @nLinhaRegistro = 0    
  end

  def IniciaBotoes()
		@btnPesquisar = Wx::Button.new(self, ID_FRMCADASTROS_BTNPESQUISAR, 
      "&Pesquisar", Wx::Point.new(248, 8), Wx::Size.new(75, 25) )
		@btnAnterior = Wx::Button.new(self, ID_FRMCADASTROS_BTNANTERIOR, 
      "&Anterior", Wx::Point.new(248, 40), Wx::Size.new(75, 25) )
    @btnAnterior.enable(false)
		@btnProximo = Wx::Button.new(self, ID_FRMCADASTROS_BTNPROXIMO, 
      "Próxi&mo", Wx::Point.new(248, 72), Wx::Size.new(75, 25) )
    @btnProximo.enable(false)
		@btnSalvar = Wx::Button.new(self, ID_FRMCADASTROS_BTNSALVAR, 
      "&Salvar", Wx::Point.new(8, 104), Wx::Size.new(75, 25) )
    @btnSalvar.enable(false)
		@btnExcluir = Wx::Button.new(self, ID_FRMCADASTROS_BTNEXCLUIR, 
      "&Excluir", Wx::Point.new(88, 104), Wx::Size.new(75, 25) )
    @btnExcluir.enable(false)
		@btnLimpar = Wx::Button.new(self, ID_FRMCADASTROS_BTNLIMPAR, 
      "&Limpar", Wx::Point.new(168, 104), Wx::Size.new(75, 25) )
		@btnFechar = Wx::Button.new(self, ID_FRMCADASTROS_BTNFECHAR, 
      "&Fechar", Wx::Point.new(248, 104), Wx::Size.new(75, 25) )
    self.btnLimparClick()
  end

  def FazLayout()
    sizer_1 = Wx::BoxSizer.new(Wx::VERTICAL)
    sizer_2 = Wx::BoxSizer.new(Wx::HORIZONTAL)
    sizer_3 = Wx::BoxSizer.new(Wx::HORIZONTAL)
    sizer_4 = Wx::BoxSizer.new(Wx::VERTICAL)
    sizer_5 = Wx::BoxSizer.new(Wx::VERTICAL)
    grid_sizer_2 = Wx::GridSizer.new(2, 1, 0, 0)
    grid_sizer_1 = Wx::GridSizer.new(2, 2, 0, 1)
    grid_sizer_1.add(@labCodigo, 0, Wx::ALIGN_BOTTOM, 0)
    grid_sizer_1.add(20, 10, 0, 0, 0)
    grid_sizer_1.add(@edtCodigo, 0, 0, 0)
    grid_sizer_1.add(@btnNovo, 0, Wx::DIRECTION_LEFT, 6)
    sizer_5.add(grid_sizer_1, 0, 0, 0)
    grid_sizer_2.add(@labDescricao, 0, Wx::ALIGN_BOTTOM, 0)
    grid_sizer_2.add(@edtDescricao, 0, 0, 2)
    sizer_5.add(grid_sizer_2, 0, 0, 0)
    sizer_3.add(sizer_5, 1, 0, 0)
    sizer_4.add(@btnPesquisar, 0, Wx::ALL|Wx::ALIGN_RIGHT, 2)
    sizer_4.add(@btnAnterior, 0, Wx::ALL|Wx::ALIGN_RIGHT, 2)
    sizer_4.add(@btnProximo, 0, Wx::ALL|Wx::ALIGN_RIGHT, 2)
    sizer_3.add(sizer_4, 0, Wx::ALIGN_RIGHT, 0)
    sizer_1.add(sizer_3, 1, Wx::ALL|Wx::EXPAND, 4)
    sizer_2.add(@btnSalvar, 0, Wx::ALL|Wx::ALIGN_RIGHT|Wx::ALIGN_BOTTOM, 2)
    sizer_2.add(@btnExcluir, 0, Wx::ALL|Wx::ALIGN_RIGHT|Wx::ALIGN_BOTTOM, 2)
    sizer_2.add(@btnLimpar, 0, Wx::ALL|Wx::ALIGN_RIGHT|Wx::ALIGN_BOTTOM, 2)
    sizer_2.add(@btnFechar, 0, Wx::ALL|Wx::ALIGN_RIGHT|Wx::ALIGN_BOTTOM, 2)
    sizer_1.add(sizer_2, 0, Wx::ALL|Wx::ALIGN_RIGHT|Wx::ALIGN_BOTTOM, 4)
    self.set_auto_layout(true)
    self.set_sizer(sizer_1)
    sizer_1.fit(self)
    sizer_1.set_size_hints(self)
  end

  def InformaLimpaDados(bInformar)
    if bInformar then
      @edtCodigo.set_value(@lstRegistros.col[@nLinhaRegistro][0])
      @edtDescricao.set_value(@lstRegistros.col[@nLinhaRegistro][1])
    else
      @edtCodigo.clear()
      @edtDescricao.clear()
      @edtCodigo.enable(true)
      @edtCodigo.set_focus()
    end
  end

  def HabilitaDados(bHabilita)
    @edtDescricao.enable(bHabilita)
  end

  def btnNovoClick()
    @btnSalvar.enable(true)
    @btnExcluir.enable(false)
    @btnPesquisar.enable(false)
    @btnAnterior.enable(false)
    @btnProximo.enable(false)
    @btnNovo.enable(false)
    
    @nLinhaRegistro = 0
    self.InformaLimpaDados(false)
    self.HabilitaDados(true)	
  end

  def NovoDado(sTextoSql)
    lstRegs = LISTAMSTR.new
  
    if $Rotinas.ConsultaDadosLista($Rotinas.gConexao, lstRegs, sTextoSql) then
      @edtCodigo.set_value(lstRegs.col[0][0])
      @edtCodigo.enable(false)
      @edtDescricao.set_focus()
    end
  end

  def btnLimparClick()
    @nLinhaRegistro = 0
    @btnSalvar.enable(false)
    @btnExcluir.enable(false)
    @btnPesquisar.enable(true)
    @btnAnterior.enable(false)
    @btnProximo.enable(false)
    @btnNovo.enable(true)
  
    self.InformaLimpaDados(false)
    self.HabilitaDados(false)	
  end

  def ExcluirDados(bValidaDados, sTextoSql)
    if bValidaDados then
      mens = Wx::MessageDialog.new(self, Resource::STR_CERTEXCL, 
        "Confirme", Wx::YES_NO | Wx::ICON_QUESTION)
      if mens.show_modal() == Wx::ID_YES then
        if $Rotinas.AtualizaDados($Rotinas.gConexao, sTextoSql) then
          self.btnLimparClick()
          return true
        end
      end
    end
    return false
  end

  def PesquisarDados(sTextoSql, sMensagemErro)
    if @edtCodigo.get_value().length == 0 then
      @btnAnterior.enable(true)
      @btnProximo.enable(true)
    end
    if $Rotinas.ConsultaDadosLista($Rotinas.gConexao, 
        @lstRegistros, sTextoSql) then
       self.InformaLimpaDados(true)
       self.HabilitaDados(true)	
  
       @btnSalvar.enable(true)
       @btnExcluir.enable(true)
       @btnPesquisar.enable(false)
       return true
    else
      $Rotinas.MessageErro(sMensagemErro)
    end  
    return false
  end

  def registroAnterior()
    if @nLinhaRegistro > 0 then
      @nLinhaRegistro = @nLinhaRegistro - 1
      self.InformaLimpaDados(true)
      return true
    else
      return false
    end
  end

  def registroProximo()
    if @nLinhaRegistro < (@lstRegistros.tam-1) then
      @nLinhaRegistro = @nLinhaRegistro + 1
      self.InformaLimpaDados(true)
      return true
    else
      return false
    end
  end

  def SalvarDados(bValidaDados, sTextoSql, 
    sTextoUpdate, sTextoInsert, bLimparDados)

    if bValidaDados then
      if $Rotinas.ConsultaDados($Rotinas.gConexao, sTextoSql) then
        sArquiva = sTextoUpdate
      else
        sArquiva = sTextoInsert
      end
      if $Rotinas.AtualizaDados($Rotinas.gConexao, sArquiva) then
        if bLimparDados then
          self.btnLimparClick()
        end
        return true
      end  
    end
    return false
  end

  def on_close(event)
    self.destroy
  end

end

